### PR TITLE
Fixed #34840 -- Avoided casting string base fields on PostgreSQL.

### PR DESCRIPTION
**Pull request by felixxm**
Thanks Alex Vandiver for the report.

Regression in 09ffc5c1212d4ced58b708cbbf3dfbfb77b782ca.

ticket-34840.

### CODE REVIEW

File: tests/constraints/tests.py [hunk 1]
```diff
@@ -995,6 +995,42 @@ def test_validate_expression_str(self):
             exclude={"name"},
         )
 
+    def test_validate_nullable_textfield_with_isnull_true(self):
+        is_null_constraint = models.UniqueConstraint(
+            "price",
+            "discounted_price",
+            condition=models.Q(unit__isnull=True),
```
Comment (apollo13) [inline tests/constraints/tests.py:1002]: Would it make sense to have a constraint that uses `Q() | Q()` to see if it propagates through properly?

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/constraints/tests.py (hunk 1 above)
Comment (charettes) [reply tests/constraints/tests.py:1002]: If we want to test `sql.where.WhereNode.as_sql` [specificity](https://github.com/django/django/blob/34ca1d3ab45f8f3f3d5bb867e4d13ca663deaa89/django/db/models/sql/where.py#L115-L187) I think we should do it directly and not through `Constraint` usage which is just one of the many feature that relies on this behaviour.

Comment (felixxm) [reply tests/constraints/tests.py:1002]: > Would it make sense to have a constraint that uses `Q() | Q()` to see if it propagates through properly?

This should be already covered by:
https://github.com/django/django/blob/34ca1d3ab45f8f3f3d5bb867e4d13ca663deaa89/tests/constraints/models.py#L22-L26

Comment (apollo13) [reply tests/constraints/tests.py:1002]: Indeed, thanks!

**Review by apollo13 — APPROVED** [review]

**Review by charettes — APPROVED** [review]
Thanks for leading the charge here @felixxm!

**Comment by felixxm** [comment]
Thanks both for checking :gift:

File: tests/constraints/tests.py [hunk 2]
```diff
@@ -995,6 +995,42 @@ def test_validate_expression_str(self):
             exclude={"name"},
         )
 
+    def test_validate_nullable_textfield_with_isnull_true(self):
```
Comment (timgraham) [inline tests/constraints/tests.py:998]: Does this test need `@skipUnlessDBFeature("supports_table_check_constraints")`? It's a required_db_feature for the `Product` table.

Comment (felixxm) [reply tests/constraints/tests.py:998]: Good catch :dart:

Comment (felixxm) [reply tests/constraints/tests.py:998]: Check out #17356.

**Pull request merged**